Motif multithread fasta input

jdimatteo opened the pull request:
Still need some refinement before merging, but I think this may be good enough fasta performance for now.

This is about 4.32x faster than the prior single threaded execution on a 4 core machine. (with hyper-threading so 8 virtual cores total).  So if the prior version was 8.3x faster than fimo, this should be 35.8x faster than fimo now.  Brings execution of a test case fasta input for chromosome 1 from ~50 seconds for fimo to 1.94 seconds now.  We need to start testing with multiple motifs or something to get longer runtimes to justify further performance improvements.

After about 4 scoring threads, performance doesn't improve much I think because the (sequential) writing of the output starts to dominate.  Even if scoring was instantaneous, we would still be at about 0.588079 seconds (~85x faster than fimo) because of the sequential file reading/writing -- again we need fasta test cases with longer execution before we start worrying about this.  Here are some performance timings/charts:

https://docs.google.com/a/boulderlabs.com/spreadsheets/d/1H9ZY6HLPY1qKIiIbbld9Pz05N0eFyY1nsoK42XWtR4g/edit?usp=sharing

jdimatteo committed:
Moved fasta processing to separate file

jdimatteo committed:
multithreaded fasta motif scoring

jdimatteo commented on liquidator/detail/score_matrix_detail.h:
```diff
@@ -79,10 +79,10 @@ scale(const PWM& pwm, const std::pair<double, double>& min_max, const unsigned r
-               const std::string& sequence,
-               const size_t begin,
-               const size_t end)
+inline unsigned score(const std::vector<std::array<unsigned, AlphabetSize>>& matrix,
```
Note that I inline all this stuff just so I don't get duplicate definition linker errors.  (Maybe I'm just lazy; wherever I can I like to avoid having a declaration separate from implementation.)  I trust compiler to inline what it sees fit regardless of what I declare inline.

adamtorgerson replied:
Not a big deal if it's working, but my understanding is that inline is only ever a hint, so this may not be portable. Making them static inline should guarantee it builds everywhere.

jdimatteo replied:
Here is some interesting discussion (with references to the standard) on inline vs static inline:

http://stackoverflow.com/a/4193698/1007353
http://stackoverflow.com/a/189162/1007353

adamtorgerson replied:
First page appears to back me up, external linkage is what we're dealing with. Second page is mixing up static functions and static variables.

Commit:
Put timing print statements in ifdefs

Commit:
Updated comments and added explicit cast to silence compiler warning

jdimatteo commented on liquidator/fasta_scorer.cpp:
```diff
@@ -0,0 +1,292 @@
+                     end <= sequence_end;
+                     ++begin, ++end)
+                {
+                    // todo: experiment with switching loop order: move matrix loop inside sequence loop
```
Actually we can't easily switch loop order, since sequence substring length depends on matrix length.

jdimatteo replied:
When two matrices of the same size (this is the case when single motif, since both forward and reverse matrices), reversing the loop order slightly decreased scoring performance by about 5%.

Commit:
multithreaded formatting output to stringstream

on 4 physical cores (hyperthreading with 8 virtual), now 6 times faster than
before mulithreading and 49.84x faster than fimo

jdimatteo commented:
I'm getting 49x faster than fimo on bb8.  On deathstar we are getting 66x faster than fimo.

I think if we had a 16 core machine, we might get 80-100x speedup.  I could spin up a c3.8xlarge aws ec2 instance with 16 ivy bridge cores (32 with hyperthreading), but I'm not going to bother right now.  With the current performance, this test case is absurdly small; on deathstar the total time takes 0.62 seconds, and so reading the input serially in 0.12 seconds is about 20% of total time.  We should probably shift focus to bams at this point anyway.
